src/common/utxobased/db/makeProcessor.ts

```diff
@@ -193,47 +186,83 @@ export async function makeProcessor(
     },
 
     async removeUtxos(utxoIds: string[]): Promise<void> {
       return await baselets.utxo(async tables => {
-        const utxos = await tables.utxoById.query('', utxoIds)
-        await tables.utxoIdsByScriptPubkey.delete(
-          '',
-          utxos.map(utxo => (utxo != null ? utxo.scriptPubkey : undefined))
-        )
+        if (utxoIds.length === 0) return
+
+        // To remove utxo from utxoIdsByScriptPubkey table:
+        // 1. Create a map of scriptPubkeys to utxoIds
+        // 2. Use the map to remove utxoIds from the utxoIdsByScriptPubkey table
+        const utxos = (await tables.utxoById.query('', utxoIds)).filter(
+          utxo => utxo != null
+        ) as IUTXO[]
+        const utxoIdsMap: { [scriptPubkey: string]: string[] } = {}
+        for (const utxo of utxos) {
+          utxoIdsMap[utxo.scriptPubkey] = [
+            ...(utxoIdsMap[utxo.scriptPubkey] ?? []),
+            utxo.id
+          ]
+        }
+        for (const scriptPubkey of Object.keys(utxoIdsMap)) {
+          const utxoIdsToRemove = utxoIdsMap[scriptPubkey]
+          const [utxoIds = []] = await tables.utxoIdsByScriptPubkey.query('', [
+            scriptPubkey
+          ])
+
+          for (const utxoIdToRemove of utxoIdsToRemove) {
+            utxoIds.splice(utxoIds.indexOf(utxoIdToRemove), 1)
+          }
+
+          // Update utxoIds for entry in table, otherwise delete entry
+          if (utxoIds.length > 0) {
+            await tables.utxoIdsByScriptPubkey.insert('', scriptPubkey, utxoIds)
+          } else {
+            await tables.utxoIdsByScriptPubkey.delete('', [scriptPubkey])
+          }
+        }
+
+        // Remove utxo utxoById table
         await tables.utxoById.delete('', utxoIds)
       })
     },
 
-    async fetchUtxos(args): Promise<IUTXO[]> {
+    async fetchUtxos(args): Promise<Array<IUTXO | undefined>> {
       const { scriptPubkey, utxoIds = [] } = args
       return await baselets.utxo(async tables => {
         if (scriptPubkey != null) {
-          const utxoIdsByScriptPubkey = (
-            await tables.utxoIdsByScriptPubkey.query('', [scriptPubkey])
-          ).filter(utxoId => utxoId != null)
+          const [
+            utxoIdsByScriptPubkey
+          ] = await tables.utxoIdsByScriptPubkey.query('', [scriptPubkey])
 
-          utxoIds.push(...utxoIdsByScriptPubkey)
+          if (utxoIdsByScriptPubkey != null)
+            utxoIds.push(...utxoIdsByScriptPubkey)
 
-          // Return undefined as the UTXO if no utxoIds are found by scriptPubkey
+          // Exit early if no utxoIds are found by scriptPubkey
           if (utxoIds.length === 0) {
-            return [undefined]
+            return []
           }
         }
 
         // Return all UTXOs if no UTXO ids are specified
         if (utxoIds.length === 0) {
-          const dump = await tables.utxoById.dumpData('')
-          return Object.values(dump.data)
+          const { data } = await tables.utxoById.dumpData('')
```

Comment:
Do not use the "dumpData" function for anything other than dumping data.
You should not expect that the return types from this function to not change.